R/S3_glance.R

```diff
@@ -0,0 +1,195 @@
+#'   \item{p.value.overid}{p-value from the chi-squared test; only reported with overidentification}
+#'
+#' @inheritParams glance.lm_robust
+#'
+#' @export
+#' @family estimatr glancers
+#' @md
+glance.iv_robust <- function(x, ...) {
+
+  if (length(x[["outcome"]]) > 1) {
+    stop("Cannot use `glance` on linear models with multiple responses.")
+  }
+
+  ret <- cbind(
+    data.frame(
+      r.squared = x[["r.squared"]],
+      adj.r.squared = x[["adj.r.squared"]],
+      df.residual = x[["df.residual"]],
+      N = as.integer(x[["N"]]),
+      se_type = x[["se_type"]]
+    ),
+    if (exists("fstatistic", x)) {
+      data.frame(
+        statistic = x[["fstatistic"]][1],
+        p.value = pf(x[["fstatistic"]][1], x[["fstatistic"]][2], x[["fstatistic"]][3], lower.tail = FALSE)
+      )
+    } else {
+      data.frame(statistic = NA_real_, p.value = NA_real_)
+    },
+    if (exists("diagnostic_firststage_fstatistic", x) && length(x[["diagnostic_firststage_fstatistic"]] == 4)) {
+      data.frame(
+        statistic.weakinst = x[["diagnostic_firststage_fstatistic"]]["value"],
+        p.value.weakinst = x[["diagnostic_firststage_fstatistic"]]["p.value"]
+      )
+    } else {
+      data.frame(statistic.weakinst = NA_real_, p.value.weakinst = NA_real_)
+    },
+    if (exists("diagnostic_endogeneity_fstatistic", x)) {
+      data.frame(
+        statistic.endogeneity = x[["diagnostic_endogeneity_fstatistic"]]["value"],
+        p.value.endogeneity = x[["diagnostic_endogeneity_fstatistic"]]["p.value"]
+      )
+    } else {
+      data.frame(statistic.endogeneity = NA_real_, p.value.endogeneity = NA_real_)
+    },
+    if (exists("diagnostic_overid_fstatistic", x)) {
+      data.frame(
+        statistic.overid = x[["diagnostic_overid_fstatistic"]]["value"],
+        p.value.overid = x[["diagnostic_overid_fstatistic"]]["p.value"]
+      )
+    } else {
+      data.frame(statistic.overid = NA_real_, p.value.overid = NA_real_)
+    }
+  )
+
+  as.data.frame(ret)
+}
+
+#' @name estimatr_glancers
+#' @templateVar class difference_in_means
+#' @return For \code{glance.difference_in_means}, a data.frame with columns:
+#'   \item{design}{the design used, and therefore the estimator used}
+#'   \item{df}{the degrees of freedom}
+#'   \item{N}{the number of observations used}
+#'   \item{N_blocks}{the number of blocks, if used}
+#'   \item{N_clusters}{the number of clusters, if used}
+#'   \item{condition2}{the second, "treatment", condition}
+#'   \item{condition1}{the first, "control", condition}
+#'
+#' @inheritParams glance.lm_robust
+#'
+#' @export
+#' @family estimatr glancers
+#' @md
+glance.difference_in_means <- function(x, ...) {
+  ret <- cbind(
+    data.frame(
+      design = x[["design"]],
+      df = x[["df"]],
+      N = as.integer(x[["N"]])
+    ),
+    if (exists("N_blocks", x)) {
```

Comment:
For these, I'd recommend wither using the coalesce operator `%||%` or a helper:

    value <- function(x. what) if(exists(what, x)) x[[what]] else NA_real_

and then instead of a cbind with if/elses, you have one data frame with

    data.frame(..., 
      N_blocks = x[["N_blocks"]] %||% NA_real_,
      ...
    )

Reply:
Thanks, will change. I'd rather just use base R so I'll stick to the function.